meson.build

```diff
@@ -0,0 +1,45 @@
+# Much of this is from SciPy
+
+project(
+  'pyoptsparse',
+  'c', 'cpp',
+  # Note that the git commit hash cannot be added dynamically here (it is added
```

Comment:
> I think the philosophy of meson, while it makes a lot of sense for compiled projects, doesn't work too well with Python. Having to list individual Python files manually is pretty error-prone, and I don't think any of the maintainers can guarantee that this installation method is going to be maintained going forward.

I quote this comment because it's related. This is technically a compiled project and the new Python module in Meson is specifically for dealing with packages that require extension modules. I'd argue that `setuptools` doesn't make sense for this project *because* of the compiled extension modules. While not available yet, with Scipy using meson as their build system, we will see a PEP517 interface that allows one to set `meson` as the build backend in the `pyproject.toml` so `pip` works naturally. It's on their list of things to do mentioned in this [article](https://scipy.github.io/devdocs/dev/contributor/meson.html). Ideally, at some point, `setuptools` will be dropped from the build process because right now this PR has a build backend calling a build backend which is not great. Once this happens, `meson` will be the one providing metadata to pip, not `setuptools`.

In summary, things are weird right now, but there should come a day soon where meson can be a replacement for `setuptools` as a backend. This means maintaining 
```
meson setup build && meson -C build install
```
and maintaining `pip install .` are one in the same. No extra work.

That said, using `meson` as a backend isn't supported yet so we could revisit this again once in the future and instead just use meson for the compiled extension modules for now. But I just wanted to get it out there that the Python community formerly using `numpy.distutils` is trending towards `meson` as a build backend which means the workflow of `pip` remains the same, it's just a different backend.

Reply:
[1f56a44](https://github.com/mdolab/pyoptsparse/pull/300/commits/1f56a444c5b7f68574e685f3ee149707a4eab42c) attempts to resolve this. I've removed(commented out) everything from `meson.build` whose functionality is currently duplicated by `setuptools` to help with maintainability on your end until `meson` gets a PEP517 interface!

Reply:
Yeah that's fair, and I can see us ditching pip if meson works that way. I'm not super familiar with all the python build/packaging options out there and the ecosystem is a bit of a mess. Since all the maintainers here are just PhD students, I really want to make sure that whatever we do accept is 1) understandable and maintainable by us, and 2) useful for everyone, but with priority given to those in the lab and our collaborators. That's why I'm pretty sold on switching from `numpy.distutils` to `meson`, but less so for `conda` and Windows builds, since our ecosystem doesn't really rely on those. In any case, thanks for being patient with my questions, and it is nice to know that people are not only using our FOSS but also willing to contribute to it.

Reply:
As a Master's student myself, and soon to be PhD student, I can empathize there haha. Just to be clear `pip` would still be the frontend for you all. I think @eli-schwartz may have linked it somewhere already, but I recommend reading [PEP517](https://peps.python.org/pep-0517/) if you haven't already. Essentially the standard opens up the build backend to anything that implements the standard, rather than locking `pip` down to `setuptools` and the like.

I will be continuously testing Windows builds for work, and since you're code doesn't do any crazy platform specific stuff, it should be fairly easy. If I come across problems on Windows I will open pull requests if I can, but definitely no expectation for Windows maintenance on your end haha. 

With `conda`/cross-compiling, it's a similar story. Except for the separation of `python_target`/`python_command` in `meson.build`, the actual cross-compilation burden is offset to those who want to use it. So I don't think it'll be  a maintenance burden on your end since `python_target` and `python_command` are the same unless otherwise specified. I don't expect the meson api for getting the python installation will change much, if at all. Also, important note, without the separation of `python_target`/`python_command`, your conda-forge package runs into problems. But yeah, I don't intend to add cross-compilation to your list either haha. Just a request that some flexibility, if simple, remains so that exotic use cases like cross-compilation are available to the user.

Finally, another use case of `python_target`/`python_command` that may be of more interest to you. Meson has Python as a dependency and by default, `py3 = py_mod.get_installation()` grabs the Python version used to run Meson. So consider the case that you use `meson` installed on your system using `apt` for example, but you want to build and install `pyoptsparse` for the Python version in your `pyenv`. Without separating `python_target`/`python_command`, this would not be possible because  `py3 = py_mod.get_installation()` would grab your system installation, not what's in `pyenv`. By separating them you can now do `py3_command = py_mod.get_installation()` for getting the system installation associated with `meson` and  `py3_target = py_mod.get_installation('/path/to/pyenv/bin/python')` to install to your `pyenv` installation.